### PR TITLE
Bugfix/vector subclass

### PR DESCRIPTION
**Pull request by zebengberg**
This PR fixes a few small bugs.

### Fixes

- Ensure the `Flight.fuel` attribute is preserved for the `Flight.filter` method.
- Ensure the `Fleet.fl_attrs` attribute is preserved for the `Fleet.filter` method.
- Raise `ValueError` when `Flight.sort` or `Fleet.sort` is called. Both of these subclasses assume a custom sorting order that is enforced in their constructors.

### Internals

- Make `Fuel` and its subclasses `JetA`, `SAFBlend`, and `HydrogenFuel` frozen.

## Tests

- [x] QC test passes locally (`make test`)
- [x] CI tests pass

@nickmasson -- FYI, I think this is independent of our API work.

### CODE REVIEW

**Review by nickmasson — APPROVED** [review]
Cool 🚀 
Will this be released as 0.49.2?

IIUC, these are fixes that we caught in the pycontrails-api test suite, right? Would you say these also constitute bugs vis a vis pycontrails itself... (i.e. would have been good to catch in the pycontrails unittests), or are bugs in the context of pycontrails integreated in pycontrails-api?

If bugs for pycontrails itself, I'm wondering if we can move some testing from pycontrails-api -> pycontrails? If that's not practical, then maybe we need a creative workflow where pycontrails-api is used as the integration test layer for pycontrails (seems to be kind of the relationship right now).  If so, we'd want to gate releases of pycontrails until we run the pycontrails-api tests on a release-candidate of pycontrails-api using the release-candidate of pycontrails.

Flow would be something like:
1. HEAD of pycontrails is a candidate for release as pycontrails from 0.50.2 -> 0.51.0
2. pycontrails-api is updated with a PR (`/update/pycontrails-version`) pointed to develop using pycontrails==main (i.e. proposed 0.51.0) in the pipfile and lock - tests are run on the PR. tests pass. pycontrails 0.51.0 cleared for release.
3. pycontrails HEAD released as 0.51.0
4. pycontrails-api `/update/pycontrails-version` branch updated to point to now-released pinned version (pycontrails==0.51.0) and merged to develop

**Review by mlshapiro — APPROVED** [review]
Looks great! Thanks for the cleanup - some of these files may have been pre-auto-formatter.

File: pycontrails/core/fleet.py [hunk 1]
```diff
@@ -121,14 +121,28 @@ def _validate(
...
+        flight_ids = set(np.unique(self["flight_id"][mask]))
+        fl_attrs = {k: v for k, v in self.fl_attrs.items() if k in flight_ids}
+        kwargs.setdefault("fl_attrs", fl_attrs)
+
+        return super().filter(mask, copy=copy, **kwargs)
+
+    @overrides
+    def sort(self, by: str | list[str]) -> NoReturn:
+        msg = (
+            "Fleet.sort is not implemented. A Fleet instance must be sorted "
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/core/fleet.py:142]: 36a2c86a3a2da78fc9fd80a4445d77b1215d17fc

File: pycontrails/core/flight.py [hunk 2]
```diff
@@ -281,8 +281,21 @@ def __init__(
...
+    @overrides
+    def filter(
+        self: FlightType, mask: npt.NDArray[np.bool_], copy: bool = True, **kwargs: Any
+    ) -> FlightType:
+        kwargs.setdefault("fuel", self.fuel)
+        return super().filter(mask, copy=copy, **kwargs)
+
+    @overrides
+    def sort(self, by: str | list[str]) -> NoReturn:
+        msg = "Flight.sort is not implemented. A Flight instance must be sorted by 'time'."
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/core/flight.py]: 36a2c86a3a2da78fc9fd80a4445d77b1215d17fc

**Comment by zebengberg** [comment]
> Cool 🚀 Will this be released as 0.49.2?
> 
> [EDIT] -- disregard below, sorry! I take "independent of our API work" to mean that these changes aren't fixes for the failing 0.49.1 upgrade sitting in pycontrails-api"...

Sorry for the confusion! Originally I was thinking that the fixes here were necessary for work on the API, but I don't think they're actually needed.

You're absolutely right that pycontrails should have more unit tests, and that the API doubles as an integration test. Like you've pieced together, I discover the pycontrails issues fixed in this PR while debugging some other stuff in the API. Typically when this happens, I create a minimal unit test over here to reproduce the issue.

I won't create a new 0.49.2 release yet unless it turns out the API needs these updates.

**Comment by zebengberg** [comment]
> some of these files may have been pre-auto-formatter.

The formatting changes are from black 23.12. A lot of the recent black minor releases have included these sorts of breaking changes.

**Pull request merged**